### ISSUE TITLE
NodeJS function - node_modules not included in deployed ZIP

### ISSUE THREAD

**Issue by justabuzz** [apex/apex#484]
If you have a bug report please let us know the following:

  - __Operating system__: Windows 10 64bit
  - __Apex Version__: 0.10.2

Hi, it isn't included and looking at the log there's no indication it's looked at at all. It only looks at the function index.js file.

apex deploy --log-level debug
   • loading functions in functions
   • loading function in functions\turnon
   • inferring runtime         function=turnon
   • inferred runtime          function=turnon runtime=nodejs4.3
   • deploying 1 functions
   • deploying                 function=turnon
   • creating build            function=turnon
   • injecting prelude         function=turnon
   • adding env                env=map[APEX_FUNCTION_NAME:turnon LAMBDA_FUNCTION_NAME:AutoOnOffEc2_turnon] function=turnon
   • add file to zip           file=index.js function=turnon
   • created build (1.1 kB)    function=turnon
   • fetching config           function=turnon
   • config unchanged          function=turnon
   • code changed              function=turnon local=w/PK2e59u2yIea/N/+jkExiZZg+tWKsMCYWD2Zu0XyU= remote=LRiUaxiUZ4wSbUGAkQX3btow/S61O7ilco2q025GmfY=
   • updating function         function=turnon
   • updated alias current     function=turnon version=4
   • function updated          function=turnon name=AutoOnOffEc2_turnon version=4
   • cleaning 1 functions
   • conditional flush         above_duration=24h0m0s above_size=500 age=1h28m32.3213361s size=289
   • close

Please let me know if you need more information.

Thanks!

**Comment by juliangruber** [comment]
in the current design `node_modules` indeed isn't included. AFAIK the way to go is to use browserify to bundle all your code into a single file, like so: https://github.com/apex/apex/blob/master/_examples/babel-browserify/project.json#L8

**Comment by justabuzz** [comment]
Thanks @juliangruber.

I reviewed the examples and read other discussions in other issues and I understand we can either use Webpack (with external aws-sdk lib) or Browserify. I have tried now with both and I'm seeing very similar error.

**Browserify:**
apex deploy --log-level debug
   • loading functions in functions
   • loading function in functions\turnon
   • deploying 1 functions
   • deploying                 function=turnon
   • creating build            function=turnon
   • injecting prelude         function=turnon
   • hook                      command=../../node_modules/.bin/browserify --node -s default -t babelify -o main.js index.js function=turnon hook=build
   ⨯ Error: function turnon: build hook:

**Webpack:**
apex deploy --log-level debug
   • loading functions in functions
   • loading function in functions\turnon
   • deploying 1 functions
   • deploying                 function=turnon
   • creating build            function=turnon
   • injecting prelude         function=turnon
   • hook                      command=../../node_modules/.bin/webpack.cmd --config ../../webpack.config.js function=turnon hook=build
   ⨯ Error: function turnon: build hook:

Some observations:
* Notice that in both it seems that it is supposed to include details after the last colon, but that's missing.
* I have tried it with both the suggested module and also with the .cmd version of each module (I'm on Windows) - the error is the identical in both.
* I have also tried to do it without changing the Webpack example at all - follow the steps in the readme and got exactly same error when tried to deploy uppercase function.

Not sure what I missing? :)
Thanks!

**Comment by DavidTPate** [comment]
@justabuzz Interesting, I use the Babel + WebPack setup with Apex for my deployments without issue. But then again I'm running on OSX. 

Is there no additional output being generated after the `Error: function turnon: build hook:`?

My initial guess would be that it might be a pathing issue. Are you able to run the `hook` command successfully yourself via the Command Prompt on Windows? ie: `../../node_modules/.bin/webpack.cmd --config ../../webpack.config.js`

**Comment by justabuzz** [comment]
@DavidTPate I've just done some more tests and here are the results. I'm sure it's an OS thing, but not sure what exactly isn't right there.

If I go to the uppercase directory and run the following command (notice the black slashes in the path to webpack.cmd) then it works:
babel-webpack\functions\uppercase>..\..\node_modules\.bin\webpack.cmd --config ../../webpack.config.js
Hash: f76c99f014648152db84
Version: webpack 1.13.1
Time: 9395ms
   Asset     Size  Chunks             Chunk Names
index.js  1.61 kB       0  [emitted]  main
    + 1 hidden modules

**But** if I try to deploy uppercase I'm getting the same result as before:
babel-webpack>apex deploy uppercase --log-level debug
   • loading functions in functions
   • loading function in functions\uppercase
   • deploying 1 functions
   • deploying                 function=uppercase
   • creating build            function=uppercase
   • injecting prelude         function=uppercase
   • hook                      command=..\..\node_modules\.bin\webpack.cmd --config ../../webpack.config.js function=uppercase hook=build
   ⨯ Error: function uppercase: build hook:

I get errors when trying to run the webpack command in requester function folder - this one looks like a 3rd party issue, though I'm curious still to figure out why it fails:
babel-webpack\functions\requester>..\..\node_modules\.bin\webpack.cmd --config ../../webpack.config.js
Hash: 5015a31b09e608531166
Version: webpack 1.13.1
Time: 10901ms
   Asset    Size  Chunks             Chunk Names
index.js  311 kB       0  [emitted]  main
    + 333 hidden modules

ERROR in ./src/index.js
Module not found: Error: Cannot resolve module 'babel-runtime/regenerator' in C:\Users\David\Work\Private\AWS\Lambdas\babel-webpack\functions\requester\src
 @ ./src/index.js 7:19-55

ERROR in ./src/index.js
Module not found: Error: Cannot resolve module 'babel-runtime/core-js/promise' in C:\Users\David\Work\Private\AWS\Lambdas\babel-webpack\functions\requester\src
 @ ./src/index.js 11:15-55

ERROR in ./src/index.js
Module not found: Error: Cannot resolve module 'babel-runtime/helpers/asyncToGenerator' in C:\Users\David\Work\Private\AWS\Lambdas\babel-webpack\functions\requester\src
 @ ./src/index.js 15:25-74

Thoughts? :)

**Comment by DavidTPate** [comment]
Interesting, it looks as though there isn't support for those hook commands on Windows currently based on [this line](https://github.com/apex/apex/blob/3f3d86ffd1d84bb34026c34707abb413429e43c3/plugins/hooks/hooks.go#L60) using `sh`.

It looks like this needs to be updated to have commands vary based on OS. @tj any thoughts?

**Comment by tj** [comment]
Hmm if you don't have node_modules in .apexignore it should be included. @DavidTPate yeah we need some better solution there :s, not sure what would be best for Windows, just regular exec's with no shell I guess? That'll hurt compatibility though

**Comment by rixrix** [comment]
re: Windows

as of Apex 0.10.2 on Windows, my `build` and `clean` hooks seems to be working fine but not using `cmd.exe` tool - BUT - I'm via the shell that came with Git for Windows (Cygwin) https://git-scm.com/download/win. FYI

**Comment by DavidTPate** [comment]
That makes sense @rixrix as that ships *nix tools with it such as `sh` that is used.

**Comment by rixrix** [comment]
@DavidTPate yep pretty much -- also w/ Windows10 you can install Ubuntu Bash. So just pick your liking Cygwin or Ubuntu ;) . cheers
